backend/app/core/search.py

```diff
@@ -0,0 +1,466 @@
+import time
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Tuple
+
+from opensearchpy import OpenSearch
+from sentence_transformers import SentenceTransformer
+
+from app.db.schemas.search import (
+    OpenSearchResponseDescriptionMatch,
+    OpenSearchResponseNameMatch,
+    OpenSearchResponseMatchBase,
+    OpenSearchResponsePassageMatch,
+    SearchRequestBody,
+    SearchResponseBody,
+    SearchResponseDocument,
+    SearchResponseDocumentPassage,
+)
+
+_DEFAULT_INNER_PRODUCT_THRESHOLD: float = 70.0
+_DEFAULT_MAX_DOC_COUNT: int = 100
+_DEFAULT_MAX_PASSAGES_PER_DOC: int = 10
+_DEFAULT_N_PASSAGES_TO_SAMPLE_PER_SHARD: int = 5000
+_DEFAULT_NAME_BOOST: int = 100
+_DEFAULT_DESCRIPTION_BOOST: int = 100
+_DEFAULT_OPENSEARCH_INDEX = "navigator"
+_DEFAULT_OPENSEARCH_REQUEST_TIMEOUT = 30
+_DEFAULT_OPENSEARCH_PREFERENCE = "prototype_user"  # TODO: document what this means
+_DEFAULT_OPENSEARCH_EMBEDDING_DIM = 768
```

Comment:
This parameter isn't needed at search time as the index mapping is designed to only accept embeddings with this length. If we wanted to do a check that the query embedding length matches the length Opensearch expects in future, we'd probably query the Opensearch index to get this value and then compare it to `emb.shape[0]` where `emb = _ENCODER.encode("query")`.